### PR TITLE
feat: Allow anonymous

### LINKED ISSUE #63: Allow Anonymous Access
## 🚀 Feature Proposal

Allow anonymous access

## Motivation

Make the authorization header optional via config
I took a look at fastify-auth and plugin scopes but cannot see a way to achive this

## Example

```js
'use strict'

const fastify = require('fastify')()
const bearerAuthPlugin = require('fastify-bearer-auth')
const keys = new Set(['a-super-secret-key', 'another-super-secret-key'])

fastify.register(bearerAuthPlugin, {keys, allowAnonymous: true})

fastify.get('/foo', (req, reply) => {
  reply.send("Public Endpoint")
})

fastify.listen({port: 8000}, (err) => {
  if (err) {
    fastify.log.error(err.message)
    process.exit(1)
  }
  fastify.log.info('http://127.0.0.1:8000/foo')
})
```

Example Changes 
```js
if (!header) {
     if (allowAnonymous) {
          return next();
     }
      const noHeaderError = Error('missing authorization header')
      fastifyReq.log.error('unauthorized: %s', noHeaderError.message)
      if (contentType) fastifyRes.header('content-type', contentType)
      fastifyRes.code(401).send(errorResponse(noHeaderError))
      return
    }
```

### PR DESCRIPTION
**Pull request by cliedeman**
Resolves #63

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: plugin.js [hunk 1]
```diff
@@ -10,16 +10,20 @@ function factory (options) {
...
   if (_options.keys instanceof Set) _options.keys = Array.from(_options.keys)
-  const { keys, errorResponse, contentType, bearerType, auth } = _options
+  const { keys, errorResponse, contentType, bearerType, auth, allowAnonymous } = _options
 
   function bearerAuthHook (fastifyReq, fastifyRes, next) {
     const header = fastifyReq.raw.headers.authorization
     if (!header) {
+      if (allowAnonymous) {
+        return next()
+      }
```
Comment (jsumners) [inline plugin.js:26]: It seems to me that this just disabled the plugin. I think this is better solved by using Fastify's encapsulation context support.

Comment (cliedeman) [reply plugin.js:26]: I wouldn't say disables the plugin. But skips the validation if the Authorization header is not present

Comment (jsumners) [reply plugin.js:26]: Then I do not understand why the plugin would be registered at all. If you can skip the authorization by omitting the header, there's no reason to add the plugin to the context.

**Review by mcollina — CHANGES_REQUESTED** [review]
This must support a way to opt-in or opt-out of the default behaviour.

You can find a way to do this in https://github.com/fastify/fastify-websocket.

You can also use https://www.fastify.io/docs/latest/Routes/#config to get a value at runtime.

File: Readme.md [hunk 2]
```diff
@@ -81,6 +83,15 @@ a `{error: message}` body; no further request processing will be performed.
 [fplugin]: https://github.com/fastify/fastify/blob/master/docs/Plugins.md
 [prehook]: https://github.com/fastify/fastify/blob/master/docs/Hooks.md
 
+The authorization header can be ommited (anonymous access) if the `allowAnonymous` config option is set to true.
```
Comment (jsumners) [inline Readme.md:86]:
```suggestion
The authorization header can be ommited (anonymous access) if the `allowAnonymous` config option is set to `true`.
```

File: test/anonymous-integration.test.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  t.plan(2)
+  fastify.inject({ method: 'GET', url: '/authenticated/default' }).then(response => {
+    t.strictEqual(response.statusCode, 401)
+    t.match(JSON.parse(response.body).error, /missing authorization header/)
+  }).catch(err => {
+    t.error(err)
+  })
+})
+
+test('authenticated: override', (t) => {
```
Comment (mcollina) [inline test/anonymous-integration.test.js:42]: please use async/await instead.

File: test/anonymous-integration.test.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  t.plan(2)
+  fastify.inject({ method: 'GET', url: '/authenticated/override' }).then(response => {
+    t.strictEqual(response.statusCode, 200)
+    t.deepEqual(JSON.parse(response.body), { hello: 'world' })
+  }).catch(err => {
+    t.error(err)
+  })
+})
+
+test('allowAnonymous: missing header succeeds', (t) => {
```
Comment (mcollina) [inline test/anonymous-integration.test.js:52]: please use async/await instead.

File: test/anonymous-integration.test.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  t.plan(2)
+  fastify.inject({ method: 'GET', url: '/allowAnonymous/default' }).then(response => {
+    t.strictEqual(response.statusCode, 200)
+    t.deepEqual(JSON.parse(response.body), { hello: 'world' })
+  }).catch(err => {
+    t.error(err)
+  })
+})
+
+test('allowAnonymous: override', (t) => {
```
Comment (mcollina) [inline test/anonymous-integration.test.js:62]: please use async/await instead.

File: test/anonymous-integration.test.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  childServer.get('/allowAnonymous/default', (req, res) => {
+    res.send({ hello: 'world' })
+  })
+
+  childServer.get('/allowAnonymous/override', { config: { allowAnonymous: false } }, (req, res) => {
+    res.send({ hello: 'world' })
+  })
+})
+
+test('authenticated: missing header fails', (t) => {
```
Comment (mcollina) [inline test/anonymous-integration.test.js:32]: please use async/await instead.

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by mcollina** [comment]
Good work!

File: Readme.md [hunk 7]
```diff
@@ -56,13 +56,15 @@ sent to the client (optional)
    rejects, or throws, a HTTP status of `500` will be sent. `req` is the Fastify
    request object. If `auth` is a function, `keys` will be ignored. If `auth` is
    not a function, or `undefined`, `keys` will be used.
+* `allowAnonymous`: allow the reques to continue the chain if the header is not preset
```
Comment (mcollina) [inline Readme.md:59]:
```suggestion
* `allowAnonymous`: allow the request to continue the chain if the header is not preset
```

File: plugin.d.ts [hunk 8]
```diff
@@ -6,7 +6,8 @@ declare namespace fastifyBearerAuth {
 		auth?: (key: string, req: fastify.FastifyRequest) => boolean | Promise<boolean>,
 		errorResponse?: (err: Error) => { error: string },
 		contentType?: string,
-		bearerType?: string
+		bearerType?: string,
+		allowAnonymous?: true
 	}
```
Comment (climba03003) [inline plugin.d.ts:11]:
```suggestion
	}
	
	interface FastifyContext {
		config: {
			allowAnonymous?: boolean
		}
	}
```

You missed the route context typing.

**Review by climba03003 — CHANGES_REQUESTED** [review]

File: plugin.d.ts (hunk 8 above)
Comment (cliedeman) [reply plugin.d.ts:11]: Thanks

**Review by climba03003 — APPROVED** [review]

**Pull request closed without merging**